drivers/retention/retention_mcp.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2023, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microchip_mcp7940n_sram
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/retention.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(retention_microchip_mcp7940n, CONFIG_RETENTION_LOG_LEVEL);
+
+#define MICROCHIP_MCP7940N_SRAM_BASE_ADDRESS 0x20
+#define MICROCHIP_MCP7940N_SRAM_DEFAULT_VALUE 0x00
+
+struct microchip_mcp7940n_retention_config {
+	struct i2c_dt_spec i2c;
+	uint32_t offset;
+	uint32_t size;
+	uint8_t verification;
+	uint8_t prefix_len;
+	uint8_t prefix[];
+};
+
+static int32_t microchip_mcp7940n_retention_init(const struct device *dev)
+{
+	const struct microchip_mcp7940n_retention_config *config = dev->config;
+
+	if (!device_is_ready(config->i2c.bus)) {
+		LOG_ERR("I2C device %s is not ready", config->i2c.bus->name);
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int32_t microchip_mcp7940n_retention_is_valid(const struct device *dev, bool *valid)
+{
+	const struct microchip_mcp7940n_retention_config *config = dev->config;
+
+	if (config->prefix_len > 0) {
+/* check if prefix matched */
+	}
+
+	if (config->verification != 0) {
+/* check if checksum matches */
+	}
+
+	/* Verified */
+	return 0;
+}
+
+static int32_t microchip_mcp7940n_retention_read(const struct device *dev, size_t offset,
+						 uint8_t *buffer, size_t *size)
+{
+	const struct microchip_mcp7940n_retention_config *config = dev->config;
+	size_t i = 0;
+	int rc;
+
+	if ((offset + *size) > config->size) {
+		return -EINVAL;
+	}
+
+	while (i < *size) {
+		LOG_DBG("Read from 0x%x", (MICROCHIP_MCP7940N_SRAM_BASE_ADDRESS + offset + i));
+		rc = i2c_reg_read_byte_dt(&config->i2c,
+					  (MICROCHIP_MCP7940N_SRAM_BASE_ADDRESS + offset + i),
+					  &buffer[i]);
+
+		if (rc != 0) {
+			return rc;
+		}
+
+		++i;
+	}
+
+	return 0;
+}
+
+static int32_t microchip_mcp7940n_retention_set(const struct device *dev, size_t offset,
+						uint8_t *buffer, size_t size)
+{
+	const struct microchip_mcp7940n_retention_config *config = dev->config;
```

Comment:
Note which affects all the drivers: modifying the memory should be protected from concurrent access from multiple threads